[Woo POS] Display "Simple products only" banner in POS mode

AnirudhBhat opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11999 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR introduces a "Simple Products Only" banner to be displayed in POS mode. The banner includes an animation that fades out and shrinks vertically upon closing. Additionally, the banner's visibility is managed via shared preferences to ensure it does not repeatedly show up once dismissed by the user.

### Design
<img width="588" alt="Screenshot 2024-07-22 at 9 31 16 AM" src="https://github.com/user-attachments/assets/11887216-78a4-4649-a62a-c37b7d2e3c76">

### Design implementation
<img width="433" alt="Screenshot 2024-07-22 at 10 00 36 AM" src="https://github.com/user-attachments/assets/e2699249-ce5f-4017-8adc-720e3b22f667">





### Changes Included
1. Added info icon for the banner.
2. Created Woo POS Banner Compose screen.
3. Introduced string resources for banner title and message.
4. Implemented logic to handle simple products only preference in AppPrefs.kt.
5. Integrated banner into the Products screen with animation.
6. Added tests to verify banner behaviour.


### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

1. Verify that the banner appears correctly in POS mode.
2. Confirm that the banner's close action triggers the expected state changes.
3. Ensure that the banner does not reappear once dismissed.
4. Ensure the banner doesn't appear on error and loading screens for products.
5. Ensure the title and the message in the banner matches the design in Figma.

**NOTE:** Please note that the "Learn more" click functionality isn't implemented yet as I need to confirm the behaviour with the designer.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/10365e4f-520a-4f0f-ad80-4fbf9ba1d7d4



- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

AnirudhBhat committed:
Add info icon

This icon will be used in the Woo POS banner screen

AnirudhBhat committed:
Add Woo POS Banner compose screen

AnirudhBhat committed:
Add title and message string resources

AnirudhBhat committed:
Add logic to get and set simple products only shared preference key in AppPrefs.kt

AnirudhBhat committed:
Add property that tells whether simple products only shared prefs key is set or not

AnirudhBhat committed:
Add 8 dp bottom padding to the POS banner

AnirudhBhat committed:
Integrate POS products only banner into Products screen

AnirudhBhat committed:
Add Banner closed event

AnirudhBhat committed:
Set the simple products only banner shared prefs when the banner is closed.

AnirudhBhat committed:
Add simple products banner boolean in the products view state

AnirudhBhat committed:
Extract into method for readability

AnirudhBhat committed:
Set simple products shared prefs in coroutine using default dispatcher

AnirudhBhat committed:
Add missing AppPrefsWrapper parameter to the view model in test file

AnirudhBhat committed:
Add test to verify that simple products boolean value is set to true when banner is closed.

AnirudhBhat committed:
Add test to verify that simple products boolean value is set to true when banner is closed.

AnirudhBhat committed:
Add test to verify that simple products boolean value is set to false initially when banner is not closed.

AnirudhBhat committed:
Add animation for Simple products only banner when closed.

AnirudhBhat committed:
Fix detekt errors

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AnirudhBhat committed:
Merge branch 'refs/heads/trunk' into issue/11999-simple-products-only-banner

AnirudhBhat committed:
Add release notes

AnirudhBhat committed:
Replace with correct info icon in the banner

AnirudhBhat committed:
Added appropriate font weight and a space to banner message.

AnirudhBhat committed:
Remove unnecessary elements from the info icon svg

AnirudhBhat committed:
Add correct color tint for the info icon

AnirudhBhat committed:
Add correct font weight for banner message

AnirudhBhat committed:
Add banner state into Woo product view state

AnirudhBhat committed:
Use the new banner state to handle banner visibility logic

AnirudhBhat committed:
Use the new banner state in the products screen

AnirudhBhat committed:
Remove unused ui event

AnirudhBhat committed:
Fix test to make use of the new banner state

AnirudhBhat committed:
Add test to verify banner shown state

AnirudhBhat committed:
Add test to verify correct title is shown in the banner

AnirudhBhat committed:
Add test to verify correct message is shown in the banner

AnirudhBhat committed:
Pass banner icon into the banner state

AnirudhBhat committed:
Get the icon from the banner state in the products screen

AnirudhBhat committed:
Pass the drawer icon into the banner state

AnirudhBhat committed:
Add banner icon into the banner state

AnirudhBhat committed:
Add test to verify that correct banner icon is shown when displayed

AnirudhBhat committed:
Fix detekt errors

kidinov commented on RELEASE-NOTES.txt:
```diff
@@ -2,7 +2,7 @@
 19.7
 -----
-
+- [Internal] Woo POS: Add Simple products only banner on Products screen [https://github.com/woocommerce/woocommerce-android/pull/12095]
```
I think we may omit release notes for POS for now, as generally POS should not touch the code of the main app, and overall probably that just creates noice for the notes generation

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/dcbe1c8bbe91bd918d44032802870035ac869688

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt:
```diff
@@ -1043,6 +1044,16 @@ object AppPrefs {
             value = value
         )
 
+    var isWooPosSimpleProductsOnlyShown: Boolean
```
We want to use the main code base as little as possible, especially considering that shared preferences are sort of [outdated](https://developer.android.com/training/data-storage/shared-preferences) already:

```
Caution: DataStore is a modern data storage solution that you should use instead of SharedPreferences. It builds on Kotlin coroutines and Flow, and overcomes many of the drawbacks of SharedPreferences.
```

I'd recommend using a data store for POS

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/3406bcd12791243030c21f127b6e00642bea3c4e

kidinov replied:
All those changes should probably be reverted?

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:
```diff
@@ -95,6 +116,14 @@ class WooPosProductsViewModel @Inject constructor(
         loadingMore = false,
         reloadingProducts = false,
+        bannerState = WooPosProductsViewState.Content.BannerState(
+            isSimpleProductsOnlyBannerShown = shouldShowProductsOnlyBanner(),
```
Just one more advantage to using data store for POS - here you hit the IO on the main thread (not blocking, but still) because of the outdated Java API shared preferences have, data store we will be safe from that

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+    Box(
+        modifier = Modifier
+            .fillMaxWidth()
+            .padding(bottom = 16.dp)
```
We may want to use `toAdaptivePadding()` to better accommodate smaller screens here for all padding on this banner

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/8814ccdb71d4360a668a7ddaf2bede5e884e60b0

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+
+@Composable
+@Suppress("DestructuringDeclarationWithTooManyEntries")
+fun WooPosBanner(
```
There are some visual mismatches with the design:
* I think the fonts are not bold but 400 in figma
* The icon is 48, here with padding it's 32?
* The color of the cross look different to the design
* Distance between title and description looks smaller in the implementation
* Not sure but it looks like that the description text is smaller than in figma

AnirudhBhat replied:
I've made the changes here: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/bd31e7784cb79e0c568fb482969968f886cdea0f

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+                    ) {
+                        Icon(
+                            painterResource(id = bannerIcon),
+                            contentDescription = "Info",
```
`contentDescriptions` should be localised probably

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/d521663f00ac0f0bfe910d8f04d6ad224bd3a756

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+                        text = annotatedText,
+                        style = MaterialTheme.typography.body1,
+                        fontWeight = FontWeight.SemiBold,
+                        color = colorResource(id = R.color.color_on_surface_high),
```
Two things here:
* We should avoid using colors from the main app so it will be easier to extract it
* You use the wrong color anyway compared to the figma - title and description have the same color =)

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/cef773943ed0015fc8a99a528cde2b5ba33388bb

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+
+                    Icon(
+                        imageVector = Icons.Default.Close,
+                        contentDescription = "Close",
```
Localisation here as well

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/d521663f00ac0f0bfe910d8f04d6ad224bd3a756

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt:
```diff
@@ -108,11 +115,43 @@ private fun WooPosProductsScreen(
+                        if (productsState.bannerState?.isSimpleProductsOnlyBannerShown == false) {
+                            AnimatedVisibility(
+                                visibleState = animVisibleState,
+                                exit = fadeOut() + shrinkVertically(),
```
I believe there is no fadeout animation on the design, could you please check again?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/df73bc98eda9e953fb195d00b8ab43d2ccad5ec5

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+
+@Composable
+@Suppress("DestructuringDeclarationWithTooManyEntries")
+fun WooPosBanner(
```
Just to think about - is this having as a reusable component justified at the moment? Do we have a place where it will be reused?

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt:
```diff
@@ -108,11 +115,43 @@ private fun WooPosProductsScreen(
-                    )
+                    Column {
+                        if (productsState.bannerState?.isSimpleProductsOnlyBannerShown == false) {
+                            AnimatedVisibility(
```
What about changing the state and simplify this?

```
    data class Content(
        val products: List<WooPosProductsListItem>,
        val loadingMore: Boolean,
        val bannerState: BannerState,
        override val reloadingProducts: Boolean = false,
    ) : WooPosProductsViewState(reloadingProducts) {
        data class BannerState(
            val isSimpleProductsOnlyBannerShown: Boolean,
            @StringRes val title: Int,
            @StringRes val message: Int,
            @DrawableRes val icon: Int,
        )
    }

                      AnimatedVisibility(
                          visible = productsState.bannerState.isSimpleProductsOnlyBannerShown,
                          exit = fadeOut() + shrinkVertically(),
                      ) {
                          WooPosBanner(
                              title = stringResource(id = productsState.bannerState.title),
                              message = stringResource(id = productsState.bannerState.message),
                              bannerIcon = R.drawable.info,
                              onClose = {
                              },
                              onLearnMore = {
                              }
                          )
                      }
```

AnirudhBhat replied:
Thanks for the suggestion. When the banner close lambda is called, it triggers recomposition, which interrupts the animation. To achieve a smooth animation, I need to observe the animation state and close the banner only after the animation finishes.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt:
```diff
@@ -108,11 +115,43 @@ private fun WooPosProductsScreen(
+                                        animVisibleState.targetState = false
+                                    },
+                                    onLearnMore = {
+                                        productsState.bannerState.onLearnMoreClicked()
```
Please check how we pass the `UIEvents` to the VM: `onItemClicked, `onEndOfProductListReached` etc

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/29a5d3860c4c4bab1a769ad1917e7aafa446444a

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:
```diff
@@ -59,6 +64,22 @@ class WooPosProductsViewModel @Inject constructor(
 
+    private fun onSimpleProductsOnlyBannerClosed() {
+        viewModelScope.launch {
+            appPrefsWrapper.isWooPosSimpleProductsOnlyBannerShown = true
```
If I am not missing something, you are on the main one here, so you update it on the main thread. Also, `withContext(Dispatchers.Main) {` doesn't do anything

AnirudhBhat replied:
AFAIU, viewModelScope.launch {} by default used `Dispatchers.Default` thread. Anyway, I'm using `DataStore` now. This shouldn't be a problem.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+                                bottom.linkTo(parent.bottom)
+                            }
+                    ) {
+                        Icon(
```
Better to wrap it in `IconButton` to have a nice circular ripple effect

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/2b50a03466fd65101cb5f9a34644455303f280fc

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt:
```diff
@@ -108,11 +115,43 @@ private fun WooPosProductsScreen(
-                        onEndOfProductListReached,
-                    )
+                    Column {
+                        if (productsState.bannerState?.isSimpleProductsOnlyBannerShown == false) {
```
Shell we add the banner in a preview so it easier to debug/develop?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/4b1e0d73d0fddd0ab8a8729e5a6bc64210e17bc5

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+                        Icon(
+                            painterResource(id = bannerIcon),
+                            contentDescription = "Info",
+                            tint = colorResource(id = R.color.woo_purple_50),
```
Try to use the colors from the WooPosTheme

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/2829eff6bd2eb0f24644c8cca5db39ace508565e

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+
+                    val annotatedText = buildAnnotatedString {
+                        append(message)
+                        withStyle(style = SpanStyle(color = colorResource(id = R.color.woo_purple_50))) {
```
And here too

AnirudhBhat replied:
Done:

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,156 @@
+                    val annotatedText = buildAnnotatedString {
+                        append(message)
+                        withStyle(style = SpanStyle(color = colorResource(id = R.color.woo_purple_50))) {
+                            append(" Learn more")
```
Localisation?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/d4ac1c68de81851e9fd20876bef56bacc87aa02e

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt:
```diff
@@ -108,11 +115,43 @@ private fun WooPosProductsScreen(
-                        onItemClicked,
-                        onEndOfProductListReached,
-                    )
+                    Column {
```
As this getting bigger now, would it make sense to extract this to it's own composable similar to the rest of the states?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/b15fda97b82480d67949a4caceb934c1cbba131e

kidinov reviewed (commented):
Hey @AnirudhBhat 👋

I left a few comments, please take a look

kidinov commented:
I also noticed that the banner seems a bit wider than the product items, unlikely on the design 

<img width="95" alt="image" src="https://github.com/user-attachments/assets/5f902810-1d8f-4c87-aa65-e452498123a3">

<img width="74" alt="image" src="https://github.com/user-attachments/assets/bf17af96-99af-4c1b-9602-3abc8bcd1db3">

AnirudhBhat committed:
Remove POS related release notes

As generally POS should not touch the code of the main app, and overall probably that just creates noice for the notes generation

AnirudhBhat committed:
Use Datastore instead of shared preference to read and write simple products only banner data

AnirudhBhat committed:
Use .toAdaptivePadding extension function for all the DP's used in the banner

AnirudhBhat committed:
UI changes to the banner

1. Changed the font weight of the title and message
2. Removed the padding from icon
3. Replaced Cross icon color with appropriate one
4. Gave some padding between title and message

AnirudhBhat committed:
Remove unused imports

AnirudhBhat committed:
Localised content descriptions

AnirudhBhat committed:
Avoids using color from main app for title and message

AnirudhBhat committed:
Use localisation for Learn more string

AnirudhBhat committed:
Add preview for product screen with simple products only banner

AnirudhBhat committed:
Wrap cross icon in an icon button so that we get a ripple effect

AnirudhBhat committed:
Change banner click actions logic for consistency

AnirudhBhat committed:
Remove fadeout animation for the banner as it was not needed.

AnirudhBhat committed:
Extract Banner display logic into a function

AnirudhBhat committed:
Add enter animation for the banner

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AnirudhBhat committed:
Use appropriate color for icon and the learn more string

AnirudhBhat committed:
Remove unused import

AnirudhBhat commented:
Thanks for the review and all your feedback @kidinov I've addressed them. Please have another look.

AnirudhBhat commented:
> I also noticed that the banner seems a bit wider than the product items, unlikely on the design

I think the banner has more height than the rest of the cards. WDYT?

<img width="607" alt="Screenshot 2024-07-23 at 2 09 21 PM" src="https://github.com/user-attachments/assets/d54d9981-12a7-4e58-980f-63e0371bdf62">

AnirudhBhat committed:
Make font weight changes to make it appear as per design

AnirudhBhat committed:
Merge branch 'refs/heads/trunk' into issue/11999-simple-products-only-banner

kidinov commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/AppPrefsTest.kt:
```diff
@@ -536,4 +536,16 @@ class AppPrefsTest {
     }
+
+    @Test
+    fun givenWooPosSimpleProductsOnlyBannerClosedThenGetterReturnsTrue() {
```
Do we still need those changes?

AnirudhBhat replied:
No. Removed here: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/5bf48d2c3cac281bdbd0c593fb78354ec5297033

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefsWrapper.kt:
```diff
@@ -32,6 +32,8 @@ class AppPrefsWrapper @Inject constructor() {
 
     var isAiProductCreationSurveyDismissed by AppPrefs::isAiProductCreationSurveyDismissed
 
+    var isWooPosSimpleProductsOnlyBannerShown by AppPrefs::isWooPosSimpleProductsOnlyShown
```
Unused too

AnirudhBhat replied:
Removed: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/ce2073460165f66188f2b390e281cf5ac7f5a715

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/WooPosTheme.kt:
```diff
@@ -49,13 +50,15 @@ private val DarkCustomColors = CustomColors(
     border = Color(0xFF8D8D8D),
     success = Color(0xFF06B166),
     error = Color(0xFFBE4400),
+    onSurfaceHigh = Color(0xDEFFFFFF)
```
I would suggest to not introduce this color, but use onBackground with specic alpha

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosBanner.kt:
```diff
@@ -0,0 +1,169 @@
+            Box(
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .padding(24.dp)
```
toAdaptivePadding? And there are more places in this files where we need to do that to margins/paddings

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/bbace0a087937537d1f71a387cf6f03fc24f2403

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:
```diff
@@ -27,6 +33,24 @@ class WooPosProductsViewModel @Inject constructor(
+    private fun observeSimpleProductsOnlyBanner() {
+        viewModelScope.launch {
+            posPreferencesRepository.isSimpleProductsOnlyBannerShown.collect { isShown ->
+                withContext(Dispatchers.Main) {
```
You update state flow - can be any thread unless am I missing something

AnirudhBhat replied:
Good point: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/83112306a54a1a3cafa58316246e903783e65dac

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:
```diff
@@ -2,23 +2,29 @@ package com.woocommerce.android.ui.woopos.home.products
     private val productsDataSource: WooPosProductsDataSource,
     private val fromChildToParentEventSender: WooPosChildrenToParentEventSender,
     private val priceFormat: WooPosFormatPrice,
+    private val posPreferencesRepository: PosPreferencesRepository
```
The naming is not consistent with the rest of the val here. Maybe `preferencesRepository`?

Also `PosPreferencesRepository` ->  `WooPosPreferencesRepository`

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/2a05611c09957bd0aaf1846e8d902f07fa6f3b35

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:
```diff
@@ -27,6 +33,24 @@ class WooPosProductsViewModel @Inject constructor(
+        observeSimpleProductsOnlyBanner()
+    }
+
+    private fun observeSimpleProductsOnlyBanner() {
```
I am not sure that if this method does anything. When the value is provided the state is not content but loading, so nothing is done. When the state changes to `content` you set the current value via `shouldShowProductsOnlyBanner`

kidinov replied:
Oh it does, but I think this is really not the optimal way?

Won't it make more sense to update the state directly?

```
private fun onSimpleProductsOnlyBannerClosed() {
        viewModelScope.launch {
            posPreferencesRepository.setSimpleProductsOnlyBannerShown(true)
           _viewState.value =            _viewState.value.copy(
        }
    }
```

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/83112306a54a1a3cafa58316246e903783e65dac

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:
```diff
@@ -124,4 +165,8 @@ class WooPosProductsViewModel @Inject constructor(
     }
+
+    private suspend fun shouldShowProductsOnlyBanner(): Boolean {
+        return posPreferencesRepository.isSimpleProductsOnlyBannerShown.first()
```
Will it be ever shown? Seems like by default it false?

AnirudhBhat replied:
Set the default to true here: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/0ca030ece379613dc12ea25c0896c11b9e042337#diff-1b2ac35ae5f85151b6d181461a9750241ec8cb6b85d15dd1861f8ebb394a30e3R19

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewState.kt:
```diff
@@ -9,8 +11,16 @@ sealed class WooPosProductsViewState(
-    ) : WooPosProductsViewState(reloadingProducts)
+    ) : WooPosProductsViewState(reloadingProducts) {
+        data class BannerState(
+            val isSimpleProductsOnlyBannerShown: Boolean,
```
The naming is confusing - does this mean that it visible now or it means that it was shown and now we should not show it?

AnirudhBhat replied:
It should basically say whether the banner was dismissed by the user or not in other words, is the banner visible on screen or not. I've tried to give it meaningful name here: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/c863bebeb490162846a8b78601c1b3cf37ffcac7

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewState.kt:
```diff
@@ -9,8 +11,16 @@ sealed class WooPosProductsViewState(
     data class Content(
         val products: List<WooPosProductsListItem>,
         val loadingMore: Boolean,
+        val bannerState: BannerState? = null,
```
You have the following combination of state declarations now:

* bannerState == null
* bannerState != null && bannerState.isSimpleProductsOnlyBannerShown == true
* bannerState != null && bannerState.isSimpleProductsOnlyBannerShown == false

So, 3 combinations for two states (visible/not visible), so potential bugs

I believe that the issues with animation you had to overcome with side effects also related to the state declaration

Here is the patch that fixes the animation and the state and a couple of things, please take a look.



https://github.com/user-attachments/assets/494c3fd6-5b65-4587-bc44-26daa3fa67e9

[Fixes_for_the_state.patch](https://github.com/user-attachments/files/16363118/Fixes_for_the_state.patch)

AnirudhBhat replied:
I have refactored the banner state here: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/0ca030ece379613dc12ea25c0896c11b9e042337#diff-1b2ac35ae5f85151b6d181461a9750241ec8cb6b85d15dd1861f8ebb394a30e3R19

kidinov reviewed (commented):
@AnirudhBhat 👋

It looks like that in the latest itteration there is a bug in when to show/not to show it so it doesn't work. Please take a look into my comments

> I think the banner has more height than the rest of the cards. WDYT?

Sorry, not following. How height would affect the width? I think the issues with padding and/or elevation of the cart that's why its wider than the products items

kidinov commented:
@AnirudhBhat Also, please take a look that there is no space before "learn more."

And I think the paddings are still slightly off, the cross is too small, and the message font is small as well

<img width="573" alt="image" src="https://github.com/user-attachments/assets/d2013f95-2149-4c79-ab82-1cc7e9149b86">

<img width="828" alt="image" src="https://github.com/user-attachments/assets/2471a0d1-206f-4110-81f5-9ba5e93fddbb">

AnirudhBhat committed:
Remove unused code

AnirudhBhat committed:
Remove unused code

AnirudhBhat committed:
Use adaptive padding for margins and paddings

AnirudhBhat committed:
Renamed PosPreferencesRepository.kt for consistency

AnirudhBhat committed:
Refactor onSimpleProductsOnlyBannerClosed

AnirudhBhat committed:
Refactor banner state and rename arguments for better understanding.

AnirudhBhat committed:
Add space before learn more string

AnirudhBhat committed:
UI changes

AnirudhBhat committed:
UI changes

AnirudhBhat committed:
Adjust banner padding so that it matches with the product list items

AnirudhBhat committed:
Remove unnecessary color

AnirudhBhat commented:
>It looks like that in the latest itteration there is a bug in when to show/not to show it so it doesn't work. Please take a look into my comments

Thanks for catching this! Sorry it slipped through. This is fixed now.

>Sorry, not following. How height would affect the width? I think the issues with padding and/or elevation of the cart that's why its wider than the products items

Ah, you mean the banner width. This has been taken care of now. It was due to the card elevation.

>Also, please take a look that there is no space before "learn more."

Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/424b4d39340b17f653fe9345a7979c2c28bd93c7

>And I think the paddings are still slightly off, the cross is too small

Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/111b719471632a51d9855a2d6652dd8483b1e81b

>the message font is small as well

As we discussed on Slack, providing the weight displayed on Figma isn't giving us proper UI as per Figma designs. I'll discuss this with Joe.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/util/ext/WooPosContextExt.kt:
```diff
@@ -2,6 +2,11 @@ package com.woocommerce.android.ui.woopos.util.ext
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.preferencesDataStore
+
+val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "pos_preferences")
```
np: `woo_pos_data_store` maybe?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/64db7eb82f93575e8b9a3388d4948522e6345d74

kidinov commented:
@AnirudhBhat 

> As we discussed on Slack, providing the weight displayed on Figma isn't giving us proper UI as per Figma designs. I'll discuss this with Joe.

Weight is one thing, but the size is set in Figma properly - just hover over the "size," and it shows the size in px. Convert it to our standard MaterialDesign font styles using this table https://m2.material.io/design/typography/the-type-system.html#type-scale

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt:
```diff
@@ -123,6 +123,7 @@ object AppPrefs {
         CHA_CHING_SOUND_ISSUE_DIALOG_DISMISSED,
         TIMES_AI_PRODUCT_CREATION_SURVEY_DISPLAYED,
         AI_PRODUCT_CREATION_SURVEY_DISMISSED,
+        WOO_POS_SIMPLE_PRODUCTS_ONLY_BANNER_SHOWN,
```
I think all the changes in this file are still redundant?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/be0851a14c4d1b7d3db1d7eb77cde0d65f17dcf8

AnirudhBhat committed:
Remove unused code

kidinov reviewed (commented):
@AnirudhBhat thank you for all the improvements!

I left a couple of comments code related, please take a look

Also, I think the left and right margins from the icon are not the same now. And the clickable area around the text with "learn more" is not symmetrical (bottom padding a missing, left and right are two big). Consider setting padding to the text to all sides, and then margins to align it properly in the layout itself


https://github.com/user-attachments/assets/04b9655f-9ec9-45f4-a441-9111c8309171

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/util/datastore/WooPosPreferencesRepository.kt:
```diff
@@ -0,0 +1,27 @@
+class WooPosPreferencesRepository @Inject constructor(private val dataStore: DataStore<Preferences>) {
+
+    companion object {
+        val SIMPLE_PRODUCTS_ONLY_BANNER_VISIBILITY = booleanPreferencesKey("isSimpleProductsOnlyBannerVisible")
```
np: maybe `is_simple_products_only_banner_visible` as it looks like that we use the notation for the keys?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/c863bebeb490162846a8b78601c1b3cf37ffcac7

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/util/datastore/WooPosPreferencesRepository.kt:
```diff
@@ -0,0 +1,27 @@
+        val SIMPLE_PRODUCTS_ONLY_BANNER_VISIBILITY = booleanPreferencesKey("isSimpleProductsOnlyBannerVisible")
+    }
+
+    val isSimpleProductsOnlyBannerVisible: Flow<Boolean> = dataStore.data
```
I think the naming is still confusing. My understanding that what you store here:
`isSimpleProductsOnlyBannerWasHiddenByUser` and then by default it should be false. wdyt?

AnirudhBhat replied:
My thought was that Simple products only banner will be shown by default unless user explicitly closes it. Anyway, your suggestion makes sense as well. I've done the changes here.https://github.com/woocommerce/woocommerce-android/pull/12095/commits/c863bebeb490162846a8b78601c1b3cf37ffcac7

AnirudhBhat committed:
Change the name of the woo pos data store preference for consistency

AnirudhBhat committed:
Adjust padding surrounding the info icon

AnirudhBhat committed:
Renamed banner state's boolean argument, data store's boolean name to store value for consistency

AnirudhBhat committed:
Added test to verify data store is updated when the banner is closed.

AnirudhBhat committed:
Merge branch 'refs/heads/trunk' into issue/11999-simple-products-only-banner

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsUIEvent.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewState.kt
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModelTest.kt

AnirudhBhat committed:
Refactored banner visibility logic

AnirudhBhat committed:
Refactored banner UI to give a properly aligned clickable area for learn more text.

AnirudhBhat commented:
>I think the left and right margins from the icon are not the same now. And the clickable area around the text with "learn more" is not symmetrical (bottom padding a missing, left and right are two big). Consider setting padding to the text to all sides, and then margins to align it properly in the layout itself

This is being taken care here: https://github.com/woocommerce/woocommerce-android/pull/12095/commits/bf391b2dd72da82d397a75d86809d3b979447b53

samiuelson reviewed (approved):
Good work, @AnirudhBhat! The only thing I'd like to consider is to move `WooPosBanner` file from the `.common.composeui.component` package to `.home.products`.

AnirudhBhat committed:
Moved WooPosBanner from common package to products package
